src/main/scala/viper/carbon/modules/impls/DefaultStateModule.scala

```diff
@@ -31,8 +31,17 @@ class DefaultStateModule(val verifier: Verifier) extends StateModule {
     Assume(currentGoodState)
   }
 
-  override def preamble = {
-    Func(Identifier(isGoodState), staticStateContributions(), Bool)
+  override def preamble: Seq[Decl] = {
+    Func(Identifier(isGoodState), staticStateContributions(), Bool) ++
+    {
+      val prevState = stateModule.state
+      stateModule.replaceState(stateModule.pureState)
+      val stateExps = components flatMap (_.currentStateExps)
```

Comment:
In theory it would nice if one could just use `StateModule.currentStateContributionValues` instead (which does not invoke the state components but rather just looks at the current `StateSnapshot`). But it does not work since it always picks the "actual state contributions" and thus does not work for the pure state. Actually (as a comment there states) it also does not do the right thing for old expressions if the old state is not expressed via the current state contributions. 

If you want and have time you can try to fix `currentStateContributionValues`, but I don't think it's necessary for this PR (since it is already broken for the old state). If you don't change it, I would add a comment here saying that it might make sense to refactor this in the future using currentStateContributionValues and add a comment to currentStateContributionValues stating that it does not work for the pure state.

Interestingly, there is a commented-out line `//  def currentStateContributions: Seq[LocalVarDecl] = components flatMap (_.currentStateContributions)` in StateModule which seems to be doing something similar to the code here.

Reply:
I added a comment.